bpf: avoid SNAT tracking for overlay traffic

julianwiedmann opened the pull request:
In certain configs (basically when the node-to-node IPv4 address is also chosen as IPV4_MASQUERADE) the [masquerade code](https://github.com/cilium/cilium/blob/17ade3372c38c423837add6049fc428656c4d476/bpf/lib/nat.h#L735) in `to-netdev` decides that tunnel traffic is potentially conflicting with masqueraded traffic. We thus need to reserve the packet's source port ("track it"), and potentially even SNAT it.

Creating SNAT entries for tunnel traffic makes little sense - in particular as the replies will not be addressed to the packet's source port, but to `TUNNEL_PORT`. Doing it anyway results in pressure on the CT and NAT maps.

Improve this by marking such traffic in `to-overlay`, and then bypassing the masquerading logic in `to-netdev `accordingly. Also use the mark to detect overlay traffic for the recently introduced [encrypted-overlay feature](https://github.com/cilium/cilium/pull/31073). For wireguard we're a bit more careful, to avoid missing unmarked traffic during an upgrade.

Fixes: #26908

```release-note
Skip overlay traffic in the BPF SNAT processing, and thus reduce pressure on the BPF Connection tracking and NAT maps.
```

julianwiedmann committed:
datapath: mark to-overlay traffic

To make smarter decisions at the native-device level (in to-netdev), mark
traffic that is created by cilium's overlay interface.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: avoid setting SNAT-done mark for to-overlay

When a packet gets SNATed in to-overlay, there is no point in setting the
SNAT-done mark. Firstly the mark refers to the *inner* packet, but all
subsequent users will only see the outer headers. Also our own netfilter
rules installed by installHostTrafficMarkRule() currently clear the mark
for overlay traffic.

Free up the mark for future usage.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nat: skip SNAT in to-netdev for overlay traffic

Creating SNAT entries for our own overlay traffic makes little sense. In
particular as the replies will not be addressed to the egressing packet's
source port, but to TUNNEL_PORT.

Avoiding such SNAT tracking reduces the pressure on the CT and NAT maps.

Fixes: https://github.com/cilium/cilium/issues/26908
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

lmb commented on pkg/bpf/testdata/unreachable-tailcall.o:
What is the source of this file?

julianwiedmann replied:
I have 0 clue, but some linter said it needed to be updated. @dylandreimerink any idea?

julianwiedmann replied:
I think Timo is poking at this right now in https://github.com/cilium/cilium/pull/31644.

ldelossa replied:
Yeah, this is test data which is (currently) invalidated when the corresponding .c file's include statements change, (the included content that is.)
This should get better once the mentioned PR is merged.

julianwiedmann replied:
rebased on-top of that PR, and now we're good :+1:

julianwiedmann commented on bpf/lib/wireguard.h:
```diff
@@ -65,31 +65,20 @@ wg_maybe_redirect_to_encrypt(struct __ctx_buff *ctx)
-			}
+		if (ctx_is_overlay(ctx)) {
+			from_tunnel = true;
+			break;
```
@cilium/wireguard I have a feeling we should only change wireguard in 1.17, after we can trust that `to-overlay` *will* mark all traffic as needed. Otherwise we might have a few unmarked packets in-flight, that would no longer get handled here.

julianwiedmann replied:
Split the wireguard part up into two parts - let's add the TODO now, and then do the conversion in v1.17.

(will drop the second patch before merging, just as preview what it will look like)

julianwiedmann replied:
alright, dropped the second patch. Let's do this.

julianwiedmann committed:
bpf: use ctx_is_overlay() in encrypted-overlay path

This cleans up some raw packet parsing. And avoids false-positives when
non-Cilium endpoints send traffic to TUNNEL_PORT.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: wireguard: add a TODO to use ctx_is_overlay()

If we switch over straight away, there's potential for a short time period
where to-netdev is already recompiled (and relies on the packet mark), but
to-overlay is still producing traffic without the mark. Such traffic would
then not get encrypted.

Therefore wait until v1.17, when we can trust that the "new" to-overlay
code is in place.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root commented on bpf/lib/nodeport.h:
> Firstly the mark refers to the *inner* packet, but all
> subsequent users will only see the outer headers.

Could you elaborate? I thought it's `skb->mark`, and it's not part of packet headers. What happens when it's wrapped into VXLAN? It might be preserved, but you say we clear it in netfilter anyway. But I kind of miss the part about inner/outer headers.

julianwiedmann replied:
When `to-overlay` currently sets `MARK_MAGIC_SNAT_DONE`, it does that based on the **inner** headers. But any subsequent observer of `skb->mark` will handle the encapsulated packet - and thus expects that the `MARK_MAGIC_SNAT_DONE` refers to the state of the **outer** headers.

In that sense we're lucky that the mark currently doesn't survive netfilter, because the semantics would be really confusing.

gentoo-root replied:
I see what you mean now, thanks for the explanation!

gentoo-root commented on bpf/lib/wireguard.h:
```diff
@@ -72,6 +72,9 @@ wg_maybe_redirect_to_encrypt(struct __ctx_buff *ctx)
 		 * This also handles IPv6, as IPv6 pkts are encapsulated w/
 		 * IPv4 tunneling.
+		 *
+		 * TODO: in v1.17, we can trust that to-overlay will mark all
```
Do we have an issue to track this? A TODO in code is likely to be forgotten.

julianwiedmann replied:
yep let me park the patch in a branch, and point an issue at it :+1:

julianwiedmann replied:
https://github.com/cilium/cilium/issues/31780

ldelossa commented on bpf/lib/overloadable_skb.h:
```diff
@@ -234,6 +234,20 @@ static __always_inline bool ctx_snat_done(const struct __sk_buff *ctx)
 
+static __always_inline void ctx_set_overlay_mark(struct __sk_buff *ctx)
+{
+	ctx->mark &= ~MARK_MAGIC_HOST_MASK;
```
This maybe a bit confusing without further comment. 

You're just using `MARK_MAGIC_HOST_MASK` because it covers the two bits you want to clear for the overlay mark, right?

julianwiedmann replied:
The mark contains other information (see the big comment in `common.h`). We need to be careful about only clearing the parts that contain the "magic" (I didn't come up with the term :smile:), same as eg. `ctx_snat_done_set()`.

brb reviewed (approved):
Thanks! Glad that you didn't change the WG to use it it right away :sweat_smile:

ldelossa commented on bpf/bpf_overlay.c:
```diff
@@ -755,19 +756,22 @@ int cil_to_overlay(struct __ctx_buff *ctx)
 	cluster_id = ctx_get_cluster_id_mark(ctx);
 #endif
+
+	ctx_set_overlay_mark(ctx);
```
Just as a sanity check, this occurs **after** vxlan encap right?

julianwiedmann replied:
This **is** the encap :smile:.

`to-overlay` processes the inner packet, then `vxlan.ko` adds the outer headers. The mark survives.

ldelossa replied:
Traced this out with @julianwiedmann 

re: https://github.com/cilium/cilium/pull/31082#pullrequestreview-1980362571

This is only a scenario if the packet is going cross net-namespace. Resolving.

ldelossa commented on bpf/bpf_host.c:
```diff
@@ -1322,23 +1321,12 @@ int cil_from_host(struct __ctx_buff *ctx)
 	if (!revalidate_data(ctx, &data, &data_end, &ipv4))
 		return DROP_INVALID;
 
-	if (!vxlan_skb_is_vxlan_v4(data, data_end, ipv4, TUNNEL_PORT))
```
At first blush I was thinking that replacing this with the overlay mark check is not equivalent, since this was determing if the traffic was vxlan. 

But I guess, internally that check is just checking if the TUNNEL_PORT matches, which in and of itself doesn't confirm anything about the tunnel proto we are using. 

I think agent side, we won't flick the ENABLE_ENCRYPTED_OVERLAY flag on if tunnel-proto is not VXLAN. So this seems okay after a bit of thought.

julianwiedmann replied:
Yeah I actually think this will take you a bit further towards GENEVE support. A tiny bit less of protocol parsing, no?

ldelossa replied:
Yup, looks good to me. It's cleaner just to catch traffic off the interface.

ldelossa reviewed (approved):
Changes to the encrypted overlay bits check out to me. 

I'm having a bit of trouble tracing out where TC runs in the egress vxlan driver path. 
I'm going to assume we are setting the mark on the packet post-encap. 

If we are not, the mark will get cleaned, and I assume this isn't occuring if the PR is tested.
